### ISSUE TITLE
Fine-tune Problem

### ISSUE THREAD

**Issue by RlChen0** [deepmind/sonnet#235]
I have a pre-train model implemented in Sonnet. How do I fine-tune the model on another dataset?
For specific, how do I freezes variables or apply different learning rate for some layer, and add new layer on the top of the model?

**Comment by tomhennigan** [comment]
Hi @RlChen0, the simplest way is to only request gradients to variables you want to train (e.g. use `module.trainable_variables` instead of `tape.watched_variables()`):

```python
pretrained = snt.nets.MLP([300, 100, 10])
finetune = snt.nets.MLP([100, 10])

def forward(x):
  x = pretrained(x)
  x = tf.nn.relu(x)
  x = finetune(x)
  return x

x = tf.ones([1, 1])
with tf.GradientTape() as tape:
  y = forward(x)
  loss = tf.reduce_sum(y)

params = finetune.trainable_variables
grads = tape.gradient(loss, params)
```

If this is cumbersome (e.g. you are using a library which itself calls `tape.watched_variables()` or you don't have one module encapsulating what you want to finetune) then you can use a context manager to stop gradients being computed for specific parts of your forward pass:

```python
def forward(x):
  with snt.custom_variable_getter(tf.stop_gradient):
    x = pretrained(x)
  x = tf.nn.relu(x)
  x = finetune(x)
  return x

with tf.GradientTape() as tape:
  y = forward(x)
  loss = tf.reduce_sum(y)

params = tape.watched_variables()
grads = tape.gradient(loss, params)
```

I put an example together in this Colab: https://colab.research.google.com/gist/tomhennigan/acf921df9a7e7dbbfd431414fd408fcb/example-of-selecting-variables-for-optimization.ipynb

**Comment by RlChen0** [comment]
Perfect solution to my problem.Thanks, Tom, you always help a lot.

**Comment by RlChen0** [comment]
Hi, @tomhennigan. I'm still confused about how to apply different learning rates.
Is there a way to do something like this?

```python
pretrained = tf.saved_model.load(model_path)
finetune = snt.Linear(100, name='finetune')

opt1 = snt.optimizers.Adam(learning_rate)
opt2 = snt.optimizers.Adam(learning_rate * 10)

opt1.apply(pretained_grads, pretrained_params)
opt2.apply(finetune_grads, finetune_params)
```

**Comment by tomhennigan** [comment]
Hi @RlChen0 , sure you just need to separate the gradients for pretrain and tune, the easiest way is to do this when you call into `tape.gradient`. Here is an example: https://colab.research.google.com/gist/tomhennigan/f6bf5520995e08fe54ade0736af9443b/example-of-selecting-variables-for-optimization.ipynb#scrollTo=WmcE3abMvMW2

**Comment by RlChen0** [comment]
Hi, @tomhennigan, but pretrained load from SavedModel has no attribute 'trainable_variables'.
```python
params = tape.watched_variables()
pretrained_param = [param for param in params
         if 'pretrained' in param.name]
finetune_param = [param for param in params
         if 'finetune' in param.name]
# or finetune_params = finetune.trainable_variables
params = (pretrain_params, finetune_params)
pretrain_grads, finetune_grads = tape.gradient(loss, params)
```
does this work?

**Comment by tomhennigan** [comment]
Hi @RlChen0 , apologies it has been quite a while since I've used TF. Yes your suggestion would work, you can also query one of the traced functions on the restored model for the parameters it uses, that might be a bit more robust (e.g. it would work even if you change the names):

https://colab.research.google.com/gist/tomhennigan/43509a6d9156cf658666d6e158b557f6/example-of-selecting-variables-for-optimization.ipynb
